src/ray/core_worker/lease_policy.h

```diff
@@ -0,0 +1,98 @@
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/common/id.h"
+#include "ray/common/task/task_spec.h"
+#include "src/ray/protobuf/common.pb.h"
+
+namespace ray {
+
+struct LocalityData {
+  uint64_t object_size;
+  absl::flat_hash_set<NodeID> nodes_containing_object;
+};
+
+/// Interface for providers of locality data to the lease policy.
+class LocalityDataProviderInterface {
+ public:
+  virtual absl::optional<LocalityData> GetLocalityData(const ObjectID &object_id) = 0;
+
+  virtual ~LocalityDataProviderInterface() {}
+};
+
+/// Interface for mocking the lease policy.
+class LeasePolicyInterface {
+ public:
+  /// Get the address of the best worker node for a lease request for the provided task.
+  virtual rpc::Address GetBestNodeForTask(const TaskSpecification &spec) = 0;
+
+  virtual ~LeasePolicyInterface() {}
+};
+
+typedef std::function<absl::optional<rpc::Address>(const NodeID &node_id)>
+    NodeAddrFactory;
+
+/// Class used by the core worker to implement a locality-aware lease policy for
+/// picking a worker node for a lease request. This class is not thread-safe.
+class LocalityAwareLeasePolicy : public LeasePolicyInterface {
+ public:
+  LocalityAwareLeasePolicy(
+      std::shared_ptr<LocalityDataProviderInterface> locality_data_provider,
+      NodeAddrFactory node_addr_factory, const rpc::Address fallback_rpc_address)
+      : locality_data_provider_(locality_data_provider),
+        node_addr_factory_(node_addr_factory),
+        fallback_rpc_address_(fallback_rpc_address) {}
+
+  ~LocalityAwareLeasePolicy() {}
+
+  /// Get the address of the best worker node for a lease request for the provided task.
+  rpc::Address GetBestNodeForTask(const TaskSpecification &spec);
+
+ private:
+  /// Get the best worker node for a lease request for the provided task.
+  absl::optional<NodeID> GetBestNodeIdForTask(const TaskSpecification &spec);
+
+  /// Provider of locality data that will be used in choosing the best lessor.
+  std::shared_ptr<LocalityDataProviderInterface> locality_data_provider_;
+
+  /// Factory for building node RPC addresses given a NodeID.
+  NodeAddrFactory node_addr_factory_;
+
+  /// RPC address of fallback node (usually the local node).
+  const rpc::Address fallback_rpc_address_;
+};
+
+/// Class used by the core worker to implement a local-only lease policy for picking
+/// a worker node for a lease request. This class is not thread-safe.
+class LocalLeasePolicy : public LeasePolicyInterface {
+ public:
+  LocalLeasePolicy(const rpc::Address fallback_rpc_address)
+      : fallback_rpc_address_(fallback_rpc_address) {}
+
+  ~LocalLeasePolicy() {}
+
+  /// Get the address of the best worker node for a lease request for the provided task.
+  rpc::Address GetBestNodeForTask(const TaskSpecification &spec);
+
+ private:
+  /// RPC address of fallback node (usually the local node).
+  const rpc::Address fallback_rpc_address_;
```

Comment:
Same question as above. Isn't it always a local node? When's the case it is other nodes?

Reply:
(If it is always local, can we change the name to `local_node_rpc_address`?)